### PR TITLE
feat: adding more request Id tracking and getters

### PR DESCRIPTION
**Pull request by kupermind**
- Adding more request Id tracking and getters;
- Add virtual functions to facilitate actions with requests;
- No current mech ABI has been changed.

### CODE REVIEW

File: contracts/AgentMech.sol [hunk 1]
```diff
@@ -38,15 +38,25 @@ contract AgentMech is ERC721Mech {
     event Request(address indexed sender, uint256 requestId, bytes data);
     event PriceUpdated(uint256 price);
 
+    enum RequestStatus {
+        DoesNotExist,
+        Requested,
+        Delivered
+    }
```
Comment (kupermind) [inline contracts/AgentMech.sol:45]: Request statuses for a getter.

File: contracts/AgentMech.sol [hunk 2]
```diff
@@ -38,15 +38,25 @@ contract AgentMech is ERC721Mech {
...
+    uint256 public numTotalRequests;
 
     // Map of requests counts for corresponding addresses
-    mapping (address => uint256) public mapRequestsCounts;
+    mapping(address => uint256) public mapRequestsCounts;
     // Map of request Ids
-    mapping (uint256 => uint256[2]) public mapRequestIds;
+    mapping(uint256 => uint256[2]) public mapRequestIds;
+    // Map of request Id => sender address
+    mapping(uint256 => address) public mapRequestAddresses;
```
Comment (kupermind) [inline contracts/AgentMech.sol:59]: Store a sender of each request Id

File: contracts/AgentMech.sol [hunk 3]
```diff
@@ -68,17 +78,26 @@ contract AgentMech is ERC721Mech {
         price = _price;
     }
 
+    /// @dev Checks for the request payment.
+    /// @param amount Amount of payment in wei.
+    function _checkRequestPayment(uint256 amount) internal virtual {
+        if (amount < price) {
+            revert NotEnoughPaid(msg.value, price);
+        }
+    }
```
Comment (kupermind) [inline contracts/AgentMech.sol:87]: Virtual payment function that can be changed by other contracts.

File: contracts/AgentMech.sol [hunk 4]
```diff
@@ -96,19 +115,31 @@ contract AgentMech is ERC721Mech {
...
 
         emit Request(msg.sender, requestId, data);
     }
 
+    /// @dev Performs actions before the delivery of a request.
+    /// @param data Self-descriptive opaque data-blob.
+    /// @return requestData Data for the request processing.
+    function _preDeliver(uint256, bytes memory data) internal virtual returns (bytes memory requestData) {
+        requestData = data;
+    }
```
Comment (kupermind) [inline contracts/AgentMech.sol:129]: The agent can wrap any data along with the request data that can be parsed during the pre-deliver function. This gives more flexibility in actions needed before the request is delivered.

**Review by 77ph — APPROVED** [review]

File: contracts/AgentMech.sol [hunk 5]
```diff
@@ -68,17 +78,28 @@ contract AgentMech is ERC721Mech {
...
-    function request(bytes memory data) external payable returns (uint256 requestId) {
-        if (msg.value < price) {
+    /// @dev Performs actions before the request is posted.
+    /// @param amount Amount of payment in wei.
+    function _preRequest(uint256 amount, uint256, bytes memory) internal virtual {
+        // Check the request payment
+        if (amount < price) {
             revert NotEnoughPaid(msg.value, price);
         }
+    }
```
Comment (kupermind) [inline contracts/AgentMech.sol:88]: Virtual pre-request function that can be changed by other contracts (for the default mech it just checks for the payment).

File: contracts/AgentMech.sol [hunk 6]
```diff
@@ -96,19 +117,31 @@ contract AgentMech is ERC721Mech {
...
+        bytes memory requestData = _preDeliver(requestId, data);
+
         // Remove delivered request Id from the request Ids map
         uint256[2] memory requestIds = mapRequestIds[requestId];
-        // Check if the request Id is invalid: previous and next request Ids are zero,
-        // and the zero's element next request Id is not equal to the provided request Id
-        if (requestIds[0] == 0 && requestIds[1] == 0 && mapRequestIds[0][1] != requestId) {
+        // Check if the request Id is invalid (non existent or delivered): previous and next request Ids are zero,
+        // and the zero's element previous request Id is not equal to the provided request Id
+        if (requestIds[0] == 0 && requestIds[1] == 0 && mapRequestIds[0][0] != requestId) {
```
Comment (DavidMinarsch) [inline contracts/AgentMech.sol:144]: Why did this change?

Comment (kupermind) [reply contracts/AgentMech.sol:144]: If this condition is true, mapRequestIds[0][0] and mapRequestIds[0][1] are the same. Just decided to have the simplest layout of the array.

**Pull request merged**